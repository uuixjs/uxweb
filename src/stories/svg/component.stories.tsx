import * as React from "react";

import {
  AlignItems,
  Background,
  Column,
  CoreText,
  Display,
  FlexDirection,
  Grid,
  InjectLayout,
  JustifyContent,
  Layout,
  SVG,
  SVGAsset,
  SVGType,
  TextType,
  Tower,
  TowerChildWidth,
} from "v2";
import {
  CombinationGenerator,
  CombinationMode,
} from "../../components/combination-generator";

export default { title: SVG.displayName };

export const svgSizeWithinContainer = () => (
  <>
    This SVG should be centered horizontally and vertically and be half the size
    of the container:
    <InjectLayout
      background={Background.Alt2}
      border
      display={Display.Flex}
      alignItems={AlignItems.Center}
      justifyContent={JustifyContent.Center}
    >
      <div style={{ width: "200px", height: "200px" }}>
        <SVG asset={SVGAsset.LogoGlitch} width={100} height={100} />
      </div>
    </InjectLayout>
  </>
);

export const svgLayout = () => (
  <>
    <Layout margin={{ y: 2 }}>
      <Layout margin={{ y: 2 }}>
        <CoreText type={TextType.H4}>Default Size:</CoreText>
        <SVG type={SVGType.Brand} asset={SVGAsset.LogoGlitch} />
        <SVG type={SVGType.Brand} asset={SVGAsset.LogoTwitch} />
      </Layout>

      <Layout margin={{ y: 2 }}>
        <CoreText type={TextType.H4}>Custom Sizes:</CoreText>

        <Layout
          background={Background.Alt2}
          display={Display.InlineBlock}
          margin={1}
        >
          <SVG
            type={SVGType.Brand}
            asset={SVGAsset.LogoGlitch}
            width={10}
            height={10}
          />
        </Layout>
        <Layout
          background={Background.Alt2}
          display={Display.InlineBlock}
          margin={1}
        >
          <SVG
            type={SVGType.Brand}
            asset={SVGAsset.LogoGlitch}
            width={25}
            height={25}
          />
        </Layout>
        <Layout
          background={Background.Alt2}
          display={Display.InlineBlock}
          margin={1}
        >
          <SVG
            type={SVGType.Brand}
            asset={SVGAsset.LogoGlitch}
            width={80}
            height={80}
          />
        </Layout>
        <Layout
          background={Background.Alt2}
          display={Display.InlineBlock}
          margin={1}
        >
          <SVG
            type={SVGType.Brand}
            asset={SVGAsset.LogoGlitch}
            width={120}
            height={120}
          />
        </Layout>
        <Layout
          background={Background.Alt2}
          display={Display.InlineBlock}
          margin={1}
        >
          <SVG
            type={SVGType.Brand}
            asset={SVGAsset.LogoGlitch}
            width={50}
            height={100}
          />
        </Layout>
        <Layout
          background={Background.Alt2}
          display={Display.InlineBlock}
          margin={1}
        >
          <SVG
            type={SVGType.Brand}
            asset={SVGAsset.LogoGlitch}
            width={100}
            height={50}
          />
        </Layout>
      </Layout>

      <Layout margin={{ y: 2 }}>
        <CoreText type={TextType.H4}>
          Fill: The SVG paths below should fill the entire container and be
          centered horizontally + vertically:
        </CoreText>
        <Grid>
          <Column cols={6}>
            <Layout fullWidth fullHeight background={Background.Alt2}>
              <SVG type={SVGType.Brand} asset={SVGAsset.LogoGlitch} fill />
            </Layout>
          </Column>
          <Column cols={6}>
            <Layout fullWidth fullHeight background={Background.Alt2}>
              <SVG type={SVGType.Brand} asset={SVGAsset.LogoTwitch} fill />
            </Layout>
          </Column>
        </Grid>
      </Layout>
    </Layout>
  </>
);

export const svgIconList = () => (
  <Tower childWidth={TowerChildWidth.Small} placeholderItems={20}>
    {Object.keys(SVGAsset)
      .filter((svg) => svg.search(/^x?small|^display|logo$|glitch$/i) < 0)
      .map((icon, index) => {
        const asset: SVGAsset = SVGAsset[icon as keyof typeof SVGAsset];
        return (
          <Layout padding={0.5} key={`icon-${index}`}>
            <Layout
              background={Background.Alt}
              padding={1}
              display={Display.Flex}
              flexDirection={FlexDirection.Column}
              alignItems={AlignItems.Center}
            >
              <Layout padding={{ y: 1 }}>
                <SVG asset={asset} width={20} height={20} />
              </Layout>
              <CoreText ellipsis>{icon}</CoreText>
            </Layout>
          </Layout>
        );
      })}
  </Tower>
);

export const svgSmallIconList = () => (
  <Tower childWidth={TowerChildWidth.Small} placeholderItems={20}>
    {Object.keys(SVGAsset)
      .filter((svg) => svg.search(/^x?small|^display/i) >= 0)
      .map((icon, index) => {
        const asset: SVGAsset = SVGAsset[icon as keyof typeof SVGAsset];
        return (
          <Layout padding={0.5} key={`icon-${index}`}>
            <Layout
              background={Background.Alt}
              padding={1}
              display={Display.Flex}
              flexDirection={FlexDirection.Column}
              alignItems={AlignItems.Center}
            >
              <Layout padding={{ y: 1 }}>
                <SVG asset={asset} width={20} height={20} />
              </Layout>
              <CoreText ellipsis>{icon}</CoreText>
            </Layout>
          </Layout>
        );
      })}
  </Tower>
);

export const svgLogoList = () => (
  <Tower childWidth={TowerChildWidth.Small} placeholderItems={20}>
    {Object.keys(SVGAsset)
      .filter((svg) => svg.search(/logo|glitch/i) >= 0)
      .map((icon, index) => {
        const asset: SVGAsset = SVGAsset[icon as keyof typeof SVGAsset];
        return (
          <Layout padding={0.5} key={`icon-${index}`}>
            <Layout
              background={Background.Alt}
              padding={1}
              display={Display.Flex}
              flexDirection={FlexDirection.Column}
              alignItems={AlignItems.Center}
            >
              <Layout padding={{ y: 1 }}>
                <SVG asset={asset} width={20} height={20} />
              </Layout>
              <CoreText ellipsis>{icon}</CoreText>
            </Layout>
          </Layout>
        );
      })}
  </Tower>
);

export const svgType = () => (
  <CombinationGenerator
    mode={CombinationMode.Simple}
    fields={[
      {
        propKey: "type",
        propEnum: SVGType,
        propEnumName: "SVGType",
      },
    ]}
  >
    <SVG asset={SVGAsset.LogoGlitch} width={50} height={50} />
  </CombinationGenerator>
);
