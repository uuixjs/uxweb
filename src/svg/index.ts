export * from "./component";
export * from "./paths";
