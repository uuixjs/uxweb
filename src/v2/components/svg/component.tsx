import { SVGAsset, SVG_PATHS } from "./paths";
import {
  getAriaProps,
  getDataProps,
  styled,
  themeTokenRule,
} from "lib/ui-utils";

import { FC } from "react";
import { ThemeToken } from "lib/ui-tokens";

export enum SVGType {
  Inherit = "color-fill-current",
  Alt = "color-fill-alt",
  Alt2 = "color-fill-alt-2",
  Brand = "color-fill-brand",
  Live = "color-text-accessible-red",
  Warn = "color-fill-warn",
  Alert = "color-fill-alert",
  Success = "color-fill-success",
  Prime = "color-fill-prime",
  Placeholder = "color-background-placeholder",
  OverlayPlaceholder = "color-background-overlay-placeholder",
}

interface ScFigureProps {
  size: "100%" | undefined;
}

const ScFigure = styled.figure<ScFigureProps>`
  align-items: center;
  display: inline-flex;
  height: ${({ size }) => size};
  width: ${({ size }) => size};
`;

interface ScSvgProps {
  /**
   * Forces all values in the SVGType enum to be valid ThemeToken names
   */
  type: ThemeToken;
}

const ScSvg = styled.svg<ScSvgProps>`
  fill: ${({ type }) => themeTokenRule(type)};
`;

export interface SVGProps {
  asset: SVGAsset;
  height?: number;
  width?: number;
  /**
   * Deprecated alias for fillParent.
   *
   * @deprecated
   */
  fill?: boolean;
  /**
   * Controls whether the icon expands to fill its parent's size.
   */
  fillParent?: boolean;
  type?: SVGType;
}

export const SVG: FC<SVGProps> = (props) => {
  const fillParent = props.fillParent || props.fill;

  return (
    <ScFigure
      className="tw-svg"
      size={fillParent ? "100%" : undefined}
      {...getDataProps(props)}
      {...getAriaProps(props)}
    >
      <ScSvg
        type={props.type || SVGType.Inherit}
        width={fillParent ? "100%" : (props.width || 20) + "px"}
        height={fillParent ? "100%" : (props.height || 20) + "px"}
        version="1.1"
        viewBox={`0 0 ${SVG_PATHS[props.asset].width} ${
          SVG_PATHS[props.asset].height
        }`}
        x="0px"
        y="0px"
      >
        {SVG_PATHS[props.asset].path}
      </ScSvg>
    </ScFigure>
  );
};

SVG.displayName = "SVG";
