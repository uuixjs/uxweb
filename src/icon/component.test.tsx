import { createMountWrapperFactory } from "tachyon-test-utils";
import { SVGAsset } from "../svg";
import {
  DEFAULT_ICON_SIZE_REM,
  Icon,
  ScIconLayout,
  ScIconSVG,
} from "./component";

describe(Icon, () => {
  const setup = createMountWrapperFactory(Icon, () => ({
    asset: SVGAsset.Heart,
  }));

  it("renders an Icon component with required props", () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders an Icon component with all props", () => {
    const { wrapper } = setup({
      fillParent: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("renders an Icon component with deprecated fill prop", () => {
    const { wrapper } = setup({
      fill: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("sets container size to default rems without fill prop", () => {
    const { wrapper } = setup();
    const iconLayout = wrapper.find(ScIconLayout).find("div").first();
    expect(iconLayout).toHaveStyleRule("height", `${DEFAULT_ICON_SIZE_REM}rem`);
    expect(iconLayout).toHaveStyleRule("width", `${DEFAULT_ICON_SIZE_REM}rem`);
  });

  it("sets container size to 100% with fill prop", () => {
    const { wrapper } = setup({ fill: true });
    const iconLayout = wrapper.find(ScIconLayout).find("div").first();
    expect(iconLayout).toHaveStyleRule("height", "100%");
    expect(iconLayout).toHaveStyleRule("width", "100%");
  });

  it("sets SVG fill to currentColor", () => {
    const { wrapper } = setup();
    expect(wrapper.find(ScIconSVG)).toHaveStyleRule("fill", "currentColor");
  });
});
