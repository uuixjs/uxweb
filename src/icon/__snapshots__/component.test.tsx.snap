// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders an Icon component with all props 1`] = `
.c3 {
  padding-bottom: 100%;
}

.c1 {
  position: relative;
  width: 100%;
  overflow: hidden;
}

.c1 > *:not(.c2) {
  position: absolute;
  left: 0;
  width: 100%;
  min-height: 100%;
  top: 0;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  width: 100%;
}

.c4 {
  fill: currentColor;
}

<Icon
  asset="Heart"
  fillParent={true}
>
  <styled.div
    className="tw-icon"
    fillParent={true}
  >
    <div
      className="c0 tw-icon"
    >
      <Aspect
        align={3}
        overflow={false}
        ratio="tw-aspect--1x1"
      >
        <styled.div
          align={3}
          allowOverflow={false}
          className="tw-aspect"
        >
          <div
            className="c1 tw-aspect"
          >
            <styled.div
              ratio={1}
            >
              <div
                className="c2 c3"
              />
            </styled.div>
            <styled.svg
              height="100%"
              version="1.1"
              viewBox="0 0 20 20"
              width="100%"
              x="0px"
              y="0px"
            >
              <svg
                className="c4"
                height="100%"
                version="1.1"
                viewBox="0 0 20 20"
                width="100%"
                x="0px"
                y="0px"
              >
                <g>
                  <path
                    clipRule="evenodd"
                    d="M9.171 4.171A4 4 0 006.343 3H6a4 4 0 00-4 4v.343a4 4 0 001.172 2.829L10 17l6.828-6.828A4 4 0 0018 7.343V7a4 4 0 00-4-4h-.343a4 4 0 00-2.829 1.172L10 5l-.829-.829z"
                    fillRule="evenodd"
                  />
                </g>
              </svg>
            </styled.svg>
          </div>
        </styled.div>
      </Aspect>
    </div>
  </styled.div>
</Icon>
`;

exports[`Icon renders an Icon component with deprecated fill prop 1`] = `
.c3 {
  padding-bottom: 100%;
}

.c1 {
  position: relative;
  width: 100%;
  overflow: hidden;
}

.c1 > *:not(.c2) {
  position: absolute;
  left: 0;
  width: 100%;
  min-height: 100%;
  top: 0;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  width: 100%;
}

.c4 {
  fill: currentColor;
}

<Icon
  asset="Heart"
  fill={true}
>
  <styled.div
    className="tw-icon"
    fillParent={true}
  >
    <div
      className="c0 tw-icon"
    >
      <Aspect
        align={3}
        overflow={false}
        ratio="tw-aspect--1x1"
      >
        <styled.div
          align={3}
          allowOverflow={false}
          className="tw-aspect"
        >
          <div
            className="c1 tw-aspect"
          >
            <styled.div
              ratio={1}
            >
              <div
                className="c2 c3"
              />
            </styled.div>
            <styled.svg
              height="100%"
              version="1.1"
              viewBox="0 0 20 20"
              width="100%"
              x="0px"
              y="0px"
            >
              <svg
                className="c4"
                height="100%"
                version="1.1"
                viewBox="0 0 20 20"
                width="100%"
                x="0px"
                y="0px"
              >
                <g>
                  <path
                    clipRule="evenodd"
                    d="M9.171 4.171A4 4 0 006.343 3H6a4 4 0 00-4 4v.343a4 4 0 001.172 2.829L10 17l6.828-6.828A4 4 0 0018 7.343V7a4 4 0 00-4-4h-.343a4 4 0 00-2.829 1.172L10 5l-.829-.829z"
                    fillRule="evenodd"
                  />
                </g>
              </svg>
            </styled.svg>
          </div>
        </styled.div>
      </Aspect>
    </div>
  </styled.div>
</Icon>
`;

exports[`Icon renders an Icon component with required props 1`] = `
.c3 {
  padding-bottom: 100%;
}

.c1 {
  position: relative;
  width: 100%;
  overflow: hidden;
}

.c1 > *:not(.c2) {
  position: absolute;
  left: 0;
  width: 100%;
  min-height: 100%;
  top: 0;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 2rem;
  width: 2rem;
}

.c4 {
  fill: currentColor;
}

<Icon
  asset="Heart"
>
  <styled.div
    className="tw-icon"
  >
    <div
      className="c0 tw-icon"
    >
      <Aspect
        align={3}
        overflow={false}
        ratio="tw-aspect--1x1"
      >
        <styled.div
          align={3}
          allowOverflow={false}
          className="tw-aspect"
        >
          <div
            className="c1 tw-aspect"
          >
            <styled.div
              ratio={1}
            >
              <div
                className="c2 c3"
              />
            </styled.div>
            <styled.svg
              height="100%"
              version="1.1"
              viewBox="0 0 20 20"
              width="100%"
              x="0px"
              y="0px"
            >
              <svg
                className="c4"
                height="100%"
                version="1.1"
                viewBox="0 0 20 20"
                width="100%"
                x="0px"
                y="0px"
              >
                <g>
                  <path
                    clipRule="evenodd"
                    d="M9.171 4.171A4 4 0 006.343 3H6a4 4 0 00-4 4v.343a4 4 0 001.172 2.829L10 17l6.828-6.828A4 4 0 0018 7.343V7a4 4 0 00-4-4h-.343a4 4 0 00-2.829 1.172L10 5l-.829-.829z"
                    fillRule="evenodd"
                  />
                </g>
              </svg>
            </styled.svg>
          </div>
        </styled.div>
      </Aspect>
    </div>
  </styled.div>
</Icon>
`;
